Adds operators to get target and feature drift

zingbretsen opened the pull request:
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Adds 2 operators--`GetFeatureDriftOperator` and `GetTargetDriftOperator` so that users can check for drift after running batch scoring in their pipelines.

## Rationale
It can be useful to check for drift after running predictions. This allows users to do it from Airflow if they want to automate the process, and then trigger other events if drift was detected.

andrius-senulis reviewed (commented):
It would be easier to pull and test your changes if you create a pull request in the same repository instead of forking it.

andrius-senulis commented on datarobot_provider/operators/datarobot.py:
```diff
@@ -259,3 +261,95 @@ def execute(self, context: Dict[str, Any]) -> str:
+        # Score data
+        self.log.info(f"Getting target drift for deployment_id={self.deployment_id}")
+
+        resp = client.get(f"deployments/{self.deployment_id}/targetDrift")
+        target_drift = resp.content.decode()
+        return json.loads(target_drift)
```
I suggest we stick to using public API client's functionality where it's available - there's a `Deployment.get_target_drift()` [method](https://datarobot-public-api-client.readthedocs-hosted.com/en/v3.0.1/autodoc/api_reference.html#datarobot.models.Deployment.get_target_drift) that we can use. You can probably get all values with `target_drift.__dict__`:

```
deployment = dr.Deployment.get(self.deployment_id)
target_drift = deployment.get_target_drift()
return json.loads(target_drift.__drift__)
```

andrius-senulis commented on datarobot_provider/operators/datarobot.py:
```diff
@@ -259,3 +261,95 @@ def execute(self, context: Dict[str, Any]) -> str:
+        # Score data
+        self.log.info(f"Getting feature drift for deployment_id={self.deployment_id}")
+
+        resp = client.get(f"deployments/{self.deployment_id}/featureDrift")
+        feature_drift = resp.content.decode()
+        return json.loads(feature_drift)
```
Same suggestion as above for `deployment.get_feature_drift()`.

zingbretsen commented:
> 

I don't have permissions to push to this repository. If you add me as a collaborator, I will push branches to this repo going forward.

jbowns10 commented:
> > 
> 
> I don't have permissions to push to this repository. If you add me as a collaborator, I will push branches to this repo going forward.

@andrius-senulis Can we add @zingbretsen as a collaborator? He's active enough and would be good to help steward the repository.

andrius-senulis commented:
@zingbretsen Sorry. I've now added CFDS team with write permissions, so you should be able to push and create PRs.

zingbretsen commented:
@andrius-senulis I added the changes requested:
- Changed to use the built-in methods for target and feature drift
- Added unit tests

I also added the operators to the example dag (including modifying the tests for the example dag).

Do you want me to push these changes to a branch in this repo and open a new PR?

andrius-senulis commented:
@zingbretsen

> Do you want me to push these changes to a branch in this repo and open a new PR?

Yes, please, it will be easier to test your changes.

andrius-senulis commented on README.md:
```diff
@@ -149,6 +149,28 @@ in the `context["params"]` variable, e.g. getting a training data you would use
+
+        None
+
+
```
You may add [this link](https://datarobot-public-api-client.readthedocs-hosted.com/en/latest-release/autodoc/api_reference.html#datarobot.models.Deployment.get_feature_drift) to the docs for optional parameters.

andrius-senulis commented on datarobot_provider/example_dags/datarobot_pipeline_dag.py:
```diff
@@ -14,6 +14,8 @@
     DeployRecommendedModelOperator,
     ScorePredictionsOperator,
     TrainModelsOperator,
+    GetFeatureDriftOperator,
+    GetTargetDriftOperator,
```
wrong import order

andrius-senulis commented on tests/unit/operators/test_datarobot.py:
```diff
@@ -7,12 +7,18 @@
     DeployRecommendedModelOperator,
     ScorePredictionsOperator,
     TrainModelsOperator,
+    GetFeatureDriftOperator,
+    GetTargetDriftOperator,
+    _serialize_drift,
```
wrong import order

andrius-senulis commented on tests/unit/operators/test_datarobot.py:
```diff
@@ -131,3 +137,55 @@ def test_operator_score_predictions(mocker):
+
+def test_operator_get_target_drift(mocker, drift_details):
+    deployment_id = "deployment-id"
+    from datarobot.models.data_drift import TargetDrift
```
please move all imports to the top of the module

andrius-senulis commented on tests/unit/operators/test_datarobot.py:
```diff
@@ -131,3 +137,55 @@ def test_operator_score_predictions(mocker):
+    from datarobot.models.data_drift import TargetDrift
+
+    target_drift = TargetDrift(**drift_details)
+    expected_target_drift = _serialize_drift(TargetDrift(**drift_details))
```
It think can use `_serialize_drift(target_drift)` here

andrius-senulis commented on tests/unit/operators/test_datarobot.py:
```diff
@@ -131,3 +137,55 @@ def test_operator_score_predictions(mocker):
+    from datarobot.models.data_drift import FeatureDrift
+
+    feature_drift = FeatureDrift(**drift_details)
+    expected_feature_drift = [_serialize_drift(FeatureDrift(**drift_details))]
```
I always suggest testing with at least two items in the list.

andrius-senulis replied:
@zingbretsen Just want to stress here how important is to always test with multiple list items. I've added another item and the test failed, as there was a bug in `_serialize_drift()` where it was changing the attributes of the original objects which resulted in errors later on, as the code expected start and date to be datetime. The fix is:

```
drift_dict = drift_obj.__dict__.copy()
```

andrius-senulis commented on tests/unit/conftest.py:
```diff
@@ -30,6 +30,7 @@ def mock_airflow_connection(mocker, dr_conn_details):
             }
-        )
+        ),
+        conn_type="Default",
```
Maybe use "DataRobot" as in your other PR, for consistency?

andrius-senulis commented:
Closed as all these changes are moved and merged in https://github.com/datarobot/airflow-provider-datarobot/pull/7
